### PR TITLE
Implement the experimental Hooks proposal behind a feature flag

### PR DESCRIPTION
**Pull request by sebmarkbage**
*Hooks* let you use all React features (such as state and lifecycle) without writing a class.

* **[Learn more about Hooks](https://reactjs.org/docs/hooks-intro.html)**
* **[Provide feedback on the Hooks RFC](https://github.com/reactjs/rfcs/pull/68)**

We have presented the Hooks proposal at React Conf 2018. This pull request implements all proposed built-in Hooks behind a feature flag, both for the client and the server renderer.

This is a very early time for Hooks and this proposal is experimental. We're merging it into master now so we can iterate on it. It is **turned on only in the 16.7 alpha** release so that you can try it and provide your feedback.

We have been using this code for a month in production so while the APIs are not stable, we don't expect the implementation to contain major bugs.

**Please keep the feedback and discussion contained [in the Hooks RFC](https://github.com/reactjs/rfcs/pull/68) (and _not_ in this PR) so that everyone can discuss it in a single place without checking multiple GitHub threads.**

### CODE REVIEW

**Comment by NE-SmallTown** [comment]
> committed on 6 Sep

Oh, the last time you do this is the *Suspense*. 😄 

Great work !

File: packages/react-reconciler/src/__tests__/ReactHooks-test.internal.js [hunk 1]
```diff
@@ -0,0 +1,1960 @@
...
+        'Layout',
+        'Layout effect 1',
+      ]);
+      expect(ReactNoop.getChildren()).toEqual([
+        span('Passive'),
+        span('Layout'),
+      ]);
+    });
+
+    it('flushes passive effects even if siblings schedule a new root', () => {
```
Comment (philipp-spiess) [inline packages/react-reconciler/src/__tests__/ReactHooks-test.internal.js]: I wonder how you come up with all of those edge cases. 😉

File: packages/react/src/ReactHooks.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    if ((Context: any)._context !== undefined) {
+      const realContext = (Context: any)._context;
+      // Don't deduplicate because this legitimately causes bugs
+      // and nobody should be using this in existing code.
+      if (realContext.Consumer === Context) {
+        warning(
+          false,
+          'Calling useContext(Context.Consumer) is not supported, may cause bugs, and will be ' +
+            'removed in a future major release. Did you mean to call useContext(Context) instead?',
+        );
```
[the remark this answers is not in the document]
Comment (trueadm) [reply packages/react/src/ReactHooks.js]: There’s no reason to use that pattern anymore with hooks.

[the remark this answers is not in the document]
Comment (Andarist) [reply packages/react/src/ReactHooks.js]: @trueadm actually this could be useful for libraries wanting to hide their providers 🤔

File: packages/react-reconciler/src/__tests__/ReactHooks-test.internal.js (hunk 1 above)
Comment (gaearon) [reply packages/react-reconciler/src/__tests__/ReactHooks-test.internal.js]: Ran into bugs in product code.

File: packages/react/src/ReactHooks.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (gaearon) [reply packages/react/src/ReactHooks.js]: It's plausible one of returned values would be a Hook itself in the future.

File: packages/react-dom/src/server/ReactPartialRendererHooks.js [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+import type {ReactContext} from 'shared/ReactTypes';
+
+import invariant from 'shared/invariant';
+import warning from 'shared/warning';
+
+type BasicStateAction<S> = S | (S => S);
```
Comment (TrySound) [inline packages/react-dom/src/server/ReactPartialRendererHooks.js]: This is more sound I think. In current case flow will always refine any value and function params and return value won't be checked. I had the same problem in powerplug.
```
type BasicStateAction<S> = (S => S) | S;
```

File: packages/react-dom/src/server/ReactPartialRendererHooks.js [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+  resolveCurrentlyRenderingComponent();
+  return context._currentValue;
+}
+
+function basicStateReducer<S>(state: S, action: BasicStateAction<S>): S {
+  return typeof action === 'function' ? action(state) : action;
+}
+
+export function useState<S>(
+  initialState: S | (() => S),
```
Comment (TrySound) [inline packages/react-dom/src/server/ReactPartialRendererHooks.js]: Here the same problem. `(() => S)` is unreachable.

File: packages/react-dom/src/server/ReactPartialRendererHooks.js (hunk 3 above)
Comment (acdlite) [reply packages/react-dom/src/server/ReactPartialRendererHooks.js]: Yeah we found the same thing when testing at FB. Forgot to update in the upstream brach. :)

**Comment by acdlite** [comment]
I'm merging this PR now so that we can continue to iterate on changes to Hooks in separate PRs. The feature flag for Hooks will remain disabled (so they won't be accessible in releases) until we come to a decision on the RFC.
